node/src/config.rs

```diff
@@ -100,7 +122,335 @@ fn validate_name(s: &str) -> Result<()> {
     Ok(())
 }
 
+#[derive(Debug, Clone)]
+pub struct EthAdapterConfig {
+    pub chain_id: ChainId,
+    pub adapters: Vec<EthereumNetworkAdapter>,
+    pub call_only: Vec<EthereumNetworkAdapter>,
+    // polling interval is set per chain so if set all adapter configuration will have
+    // the same value.
+    pub polling_interval: Option<Duration>,
+}
+
+#[derive(Debug, Clone)]
+pub struct FirehoseAdapterConfig {
+    pub chain_id: ChainId,
+    pub kind: BlockchainKind,
+    pub adapters: Vec<Arc<FirehoseEndpoint>>,
+}
+
+#[derive(Debug, Clone)]
+pub enum AdapterConfiguration {
+    Rpc(EthAdapterConfig),
+    Firehose(FirehoseAdapterConfig),
+    Substreams(FirehoseAdapterConfig),
+}
+
+impl AdapterConfiguration {
+    pub fn kind(&self) -> &BlockchainKind {
+        match self {
+            AdapterConfiguration::Rpc(_) => &BlockchainKind::Ethereum,
+            AdapterConfiguration::Firehose(fh) | AdapterConfiguration::Substreams(fh) => &fh.kind,
+        }
+    }
+    pub fn chain_id(&self) -> &ChainId {
+        match self {
+            AdapterConfiguration::Rpc(EthAdapterConfig { chain_id, .. })
+            | AdapterConfiguration::Firehose(FirehoseAdapterConfig { chain_id, .. })
+            | AdapterConfiguration::Substreams(FirehoseAdapterConfig { chain_id, .. }) => chain_id,
+        }
+    }
+
+    pub fn as_rpc(&self) -> Option<&EthAdapterConfig> {
+        match self {
+            AdapterConfiguration::Rpc(rpc) => Some(rpc),
+            _ => None,
+        }
+    }
+
+    pub fn as_firehose(&self) -> Option<&FirehoseAdapterConfig> {
+        match self {
+            AdapterConfiguration::Firehose(fh) => Some(fh),
+            _ => None,
+        }
+    }
+
+    pub fn as_substreams(&self) -> Option<&FirehoseAdapterConfig> {
+        match self {
+            AdapterConfiguration::Substreams(fh) => Some(fh),
+            _ => None,
+        }
+    }
+}
+
+pub struct Networks {
+    pub adapters: Vec<AdapterConfiguration>,
+    rpc_provider_manager: ProviderManager<EthereumNetworkAdapter>,
+    firehose_provider_manager: ProviderManager<Arc<FirehoseEndpoint>>,
+    substreams_provider_manager: ProviderManager<Arc<FirehoseEndpoint>>,
+}
+
+impl Networks {
+    // noop is important for query_nodes as it shortcuts a lot of the process.
+    fn noop() -> Self {
+        Self {
+            adapters: vec![],
+            rpc_provider_manager: ProviderManager::new(
+                Logger::root(Discard, o!()),
+                vec![].into_iter(),
+                Arc::new(MockIdentValidator),
```

Comment:
Can't this be the real validator? If there's no endpoints, the validator won't have to do anything. I don't like having all these traits floating around where whether they do anything or not depends on setup and you're never quite sure if the setup did things correctly.

Reply:
I've only implemented this for BlockStore and then the different test/mock implementations, I could add a Noop implementation but otherwise we don't have the necessary information to create it without arguments. This struct is not supposed to be mutated manually so the fact it it generated as noop is just intended for situations where it is expected to be used as noop/empty. Not sure yet another empty implementation is really useful, what do you think?